File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.java


import java.util.Arrays;
import java.util.Scanner;

public class Main {

	public static void main(String[] args) {

		Scanner sc = new Scanner(System.in);

		n = sc.nextInt();
		r = n / 2;
		grid = new int[n][n];
		arr = new int[n];
		for (int i = 0; i < n; i++) {
			arr[i] = i;
		}
//		System.out.println(Arrays.toString(arr));

		sela = new int[r];
		cnt = 0;
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n; j++) {
				grid[i][j] = sc.nextInt();
			}
		}
		ans = Integer.MAX_VALUE;
		comb(0, 0);
		System.out.println(ans);

	}

	static int[][] grid;
	static int ans;
	static int cnt;
	static int n;
	static int r;
	static int sela[];
	static int selb[];
	static int arr[];
	static boolean check[];

	private static void comb(int idx, int sidx) {
		if (sidx == r) {
			selb = new int[r];
			check = new boolean[n];

			for (int i = 0; i < r; i++) {
				check[sela[i]] = true;
			}

			int bindex = 0;
			for (int i = 0; i < n; i++) {
				if (!check[i]) {
					selb[bindex++] = i;
				}
			}


			int atotal = 0;
			int btotal = 0;
			for (int i = 0; i < r; i++) {
				for (int j = i + 1; j < r; j++) {
					atotal += grid[sela[i]][sela[j]]+grid[sela[j]][sela[i]];
					btotal += grid[selb[i]][selb[j]]+grid[selb[j]][selb[i]];
				}
			}
			ans = Math.min(ans, Math.abs(atotal - btotal));

			return;
		}

		if (idx == n) {
			return;
		}
		

		sela[sidx] = arr[idx];
		comb(idx + 1, sidx + 1);
		comb(idx + 1, sidx);

	}

}

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.py
'''
n개중 m개를 뽑는 조합
1234중 12를 뽑으면 안뽑힌 34가 나머지!
1234중 34를 뽑으면 안뽑힌 12가 나머지!
이 둘은 중복되는데..... 조합을 절반만 돌리겠음
-> 그렇게 하기 위해서 start를 들고 다니면서 절반 왔으면 return 하는 조건이 필요!
'''
n = int(input())
grid = [list(map(int, input().split())) for i in range(n)]

arr = list(range(n))
m = n//2
sel = [0]* m # 절반만 선택할 것

ans = 100001

def  btk(idx,sidx,start):
    global ans
    if(start == n//2): # 더이상은 중복이야
        return

    if(sidx ==m):
        not_sel = []
        for i in range(n):
            if(i not in sel):
                not_sel.append(i)
        # print(sel,not_sel)

        sel_score = 0
        for i in range(m):
            for j in range(i+1,m):
                sel_score += grid[sel[i]][sel[j]]
                sel_score += grid[sel[j]][sel[i]]
                
        not_sel_score = 0
        for i in range(m):
            for j in range(i+1,m):
                not_sel_score += grid[not_sel[i]][not_sel[j]]
                not_sel_score += grid[not_sel[j]][not_sel[i]]
        # print(sel_score,not_sel_score)
        ans = min(ans, abs(not_sel_score-sel_score))
        return
    if(idx==n):
        return

    sel[sidx] = arr[idx]
    btk(idx+1,sidx+1,start)
    btk(idx+1,sidx,start+1)


btk(0,0,0) # start
print(ans)
